hyperparameter_tuning.py

```diff
@@ -0,0 +1,155 @@
+import gc
+
+import numpy as np
+from deepsense import neptune
+from sklearn.externals import joblib
+
+from steps.base import BaseTransformer
+from utils import set_seed
+
+
+class RandomSearchOptimizer(BaseTransformer):
+    def __init__(self, TransformerClass, params,
+                 score_func, maximize,
+                 train_input_keys, valid_input_keys,
+                 n_runs,
+                 callbacks=[]):
+        self.TransformerClass = TransformerClass
+        self.param_space = create_param_space(params, n_runs)
+        self.train_input_keys = train_input_keys
+        self.valid_input_keys = valid_input_keys
+        self.score_func = score_func
+        self.maximize = maximize
+        self.callbacks = callbacks
+        self.best_transformer = TransformerClass(**self.param_space[0])
+
+    def fit(self, **kwargs):
+        if self.train_input_keys:
+            train_inputs = {input_key: kwargs[input_key] for input_key in self.train_input_keys}
+        else:
+            train_inputs = kwargs
+        X_valid, y_valid = kwargs[self.valid_input_keys[0]], kwargs[self.valid_input_keys[1]]
+
+        results = []
+        for i, param_set in enumerate(self.param_space):
+            try:
+                transformer = self.TransformerClass(**param_set)
+                transformer.fit(**train_inputs)
+            except Exception:
+                continue
+            y_pred_valid = transformer.transform(X_valid)
+            y_pred_valid_value = list(y_pred_valid.values())[0]
+            run_score = self.score_func(y_valid, y_pred_valid_value)
+            results.append((run_score, param_set))
+
+            del y_pred_valid, transformer
+            gc.collect()
+
+            for callback in self.callbacks:
+                callback.on_run_end(score=run_score, params=param_set)
+
+        assert len(results) > 0, 'All random search runs failed, check your parameter space'
+        results_sorted = sorted(results, key=lambda x: x[0])
+
+        if self.maximize:
+            best_score, best_param_set = results_sorted[-1]
+        else:
+            best_score, best_param_set = results_sorted[0]
+
+        for callback in self.callbacks:
+            callback.on_search_end(results=results)
+
+        self.best_transformer = self.TransformerClass(**best_param_set)
```

Comment:
Why do I have to inject Transformer inside, transformer? It seems that we are breaking our abstraction, here. What do you think @pknut ?